README.md
# UNICEF Snapshot

Snapshot is a library that provides snapshot of data changes on django model.


## Installation

   pip install unicef-snapshot


## Setup

Add ``unicef_snapshot`` to ``INSTALLED_APPS`` in ``settings.py``

   INSTALLED_APPS = [
       ...
       'unicef_snapshot',
   ]


## Usage

Use ``unicef-snapshot`` in serializers

   from unicef_snapshot.serializers import SnapshotModelSerializer

   class ExampleSerializer(SnapshotModelSerializer):
       ...


## Contributing

### Environment Setup

To install the necessary libraries

   $ make install


### Coding Standards

See `PEP 8 Style Guide for Python Code <https://www.python.org/dev/peps/pep-0008/>`_ for complete details on the coding standards.

To run checks on the code to ensure code is in compliance

   $ make lint


### Testing

Testing is important and tests are located in `tests/` directory and can be run with;

   $ make test

Coverage report is viewable in `build/coverage` directory, and can be generated with;


### Links


+--------------------+----------------+--------------+--------------------+
| Stable             |                | |master-cov| |                    |
+--------------------+----------------+--------------+--------------------+
| Development        |                | |dev-cov|    |                    |
+--------------------+----------------+--------------+--------------------+
| Source Code        |https://github.com/unicef/unicef-snapshot           |
+--------------------+----------------+-----------------------------------+
| Issue tracker      |https://github.com/unicef/unicef-snapshot/issues    |
+--------------------+----------------+-----------------------------------+


CHANGES
Release 1.4 (in development)
-----------
* added support to Django 5.0
* added support to python 3.11, 3.12


Release 1.3
-----------
* added support Django 4.1
* admin performance use raw_id for users


Release 1.2
-----------
* added support Django 4.0


Release 1.1
-----------
* fixed migrations


Release 1.0
-----------
* Add support to python 3.9
* Remove support to python < 3.9
* Add support to django 3.2
* Remove support to django < 3.2


Release 0.2.3
-------------
* Remove add permission from Django admin
* Add Django 2.11 support
* Drop Django 1.11 support
* Update requirements


test.sh
#!/usr/bin/env bash
export PATH=$PATH:/home/circleci/.local/bin
pip install -r src/requirements/install.pip --user
pip install -r src/requirements/testing.pip --user
pytest tests \
  -q \
  --create-db \
  --cov-report=html \
  --cov-report=term \
  --cov-config=tests/.coveragerc \
  --cov=unicef_snapshot


tests/conftest.py
from django.contrib.admin.sites import AdminSite
from rest_framework.test import APIClient

import pytest

from tests import factories


@pytest.fixture()
def api_client():
    return APIClient()


@pytest.fixture
def admin_site():
    return AdminSite()


@pytest.fixture
def user():
    return factories.UserFactory()


@pytest.fixture
def author():
    return factories.AuthorFactory()


tests/factories.py
from django.contrib.auth import get_user_model

import factory
import factory.fuzzy
from factory import random

from unicef_snapshot import models

from demo.sample.models import Author, Book, Tag


class UserFactory(factory.django.DjangoModelFactory):
    username = factory.Faker("user_name")
    email = factory.Faker("email")

    class Meta:
        model = get_user_model()


class TagFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")

    class Meta:
        model = Tag


class AuthorFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")

    class Meta:
        model = Author


class BookFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    author = factory.SubFactory(AuthorFactory)

    class Meta:
        model = Book


class FuzzyActivityAction(factory.fuzzy.BaseFuzzyAttribute):
    def fuzz(self):
        return random.randgen.choice([a[0] for a in models.Activity.ACTION_CHOICES])


class ActivityFactory(factory.django.DjangoModelFactory):
    target = factory.SubFactory(AuthorFactory)
    action = FuzzyActivityAction()
    by_user = factory.SubFactory(UserFactory)
    data = {"random": "data"}
    change = ""

    class Meta:
        model = models.Activity


src/unicef_snapshot/admin.py
from django.contrib import admin
from django.contrib.contenttypes.admin import GenericTabularInline

from unicef_snapshot.models import Activity
from unicef_snapshot.utils import create_dict_with_relations, create_snapshot


class SnapshotModelAdmin(admin.ModelAdmin):
    def save_model(self, request, obj, form, change):
        pre_save = None
        if hasattr(obj, "pk") and obj.pk is not None:
            pre_save = obj.__class__.objects.get(pk=obj.pk)
        pre_save = create_dict_with_relations(pre_save)
        super().save_model(request, obj, form, change)
        create_snapshot(obj, pre_save, request.user)


class ActivityInline(GenericTabularInline):
    model = Activity
    ct_field = "target_content_type"
    ct_fk_field = "target_object_id"
    fields = ["action", "by_user_display", "change", "created"]
    readonly_fields = [
        "action",
        "by_user_display",
        "change",
        "created",
    ]
    extra = 0
    can_delete = False
    can_add = False

    def has_add_permission(self, request, obj=None):
        return False


class ActivityAdmin(admin.ModelAdmin):
    model = Activity
    readonly_fields = [
        "target_content_type",
        "target_object_id",
        "target",
        "action",
        "by_user_display",
        "data",
        "change",
        "created",
        "modified",
    ]
    list_display = ("target", "action", "by_user_display")
    list_filter = ("action",)
    raw_id_fields = ("by_user",)

    def has_add_permission(self, request):
        return False


admin.site.register(Activity, ActivityAdmin)


src/unicef_snapshot/utils.py
from django.db.models.query import QuerySet
from django.forms import model_to_dict

from unicef_snapshot.models import Activity


def jsonify(data):
    """Convert data into a dictionary that can be json encoded"""
    allowed_types = (
        str,
        bool,
        dict,
        int,
        float,
        list,
        set,
        tuple,
    )
    for key, value in data.items():
        if isinstance(value, QuerySet):
            data[key] = [v.pk for v in value]
        if not isinstance(value, allowed_types):
            data[key] = str(data[key])
    return data


def get_to_many_field_names(cls):
    """Get all the many_to_many and one_to_many field names for a class"""
    fields = []
    for field in cls._meta.get_fields():
        if field.one_to_many or field.many_to_many:
            fields.append(field.name)
    return fields


def create_dict_with_relations(obj):
    """Convert obj instance to a dictionary and then set
    many_to_many and one_to_many relation values, using their pk values

    These relations do not show up initially in the conversion of object
    to a dictionary format.
    """
    obj_dict = {}
    if obj is not None:
        # re-query obj to by-pass any caching (prefetch_related)
        obj = obj.__class__.objects.get(pk=obj.pk)
        obj_dict = jsonify(model_to_dict(obj))
        fields = get_to_many_field_names(obj)
        for field_name in fields:
            if hasattr(obj, field_name):
                field = getattr(obj, field_name)
                obj_dict[field_name] = [x.pk for x in field.all()]

    return obj_dict


def create_change_dict(prev_dict, current_dict):
    """Create a dictionary showing the differences between the
    initial target and the target after being saved.

    If prev_dict is empty, then change is empty as well
    """
    change = {}
    if prev_dict is not None:
        for k, v in prev_dict.items():
            if k in current_dict and current_dict[k] != prev_dict[k]:
                change.update(
                    {
                        k: jsonify(
                            {
                                "before": prev_dict[k],
                                "after": current_dict[k],
                            }
                        )
                    }
                )

    return change


def create_snapshot(target, target_before, by_user):
    """If target_before is empty, then action is create, otherwise update

    Create a dictionary of change between target before save and after.
    """
    action = Activity.UPDATE if target_before else Activity.CREATE
    current_obj_dict = create_dict_with_relations(target)
    change = create_change_dict(target_before, current_obj_dict)

    activity_kwargs = {
        "target": target,
        "by_user": by_user,
        "action": action,
        "data": current_obj_dict,
        "change": change,
    }

    snapshot_additional_data = getattr(target, "snapshot_additional_data", None)
    if callable(snapshot_additional_data):
        activity_kwargs["data"].update(snapshot_additional_data(change))

    activity = Activity.objects.create(**activity_kwargs)

    return activity


src/unicef_snapshot/views.py
import datetime

from django.contrib.contenttypes.models import ContentType
from django.utils import timezone
from rest_framework.generics import ListAPIView
from rest_framework.permissions import IsAdminUser

from unicef_snapshot.models import Activity
from unicef_snapshot.serializers import ActivitySerializer
from unicef_snapshot.utils import create_dict_with_relations, create_snapshot


class ActivityListView(ListAPIView):
    serializer_class = ActivitySerializer
    permission_claasses = (IsAdminUser,)

    def get_queryset(self):
        queryset = Activity.objects.all()
        tz = timezone.get_default_timezone()

        # filter on user
        user = self.request.query_params.get("user", None)
        if user is not None:
            queryset = queryset.filter(by_user__email=user)

        # filter on target
        target = self.request.query_params.get("target", None)
        if target is not None:
            content_types = ContentType.objects.filter(model=target.lower())
            queryset = queryset.filter(target_content_type__in=content_types)

        # filter on action
        action = self.request.query_params.get("action", None)
        if action is not None:
            queryset = queryset.filter(action=action)

        # filter on date_from (yyyy-mm-dd)
        date_from = self.request.query_params.get("date_from", None)
        if date_from is not None:
            try:
                date_from = datetime.datetime.strptime(date_from, "%Y-%m-%d")
                date_from = timezone.make_aware(date_from, tz)
            except ValueError:
                # return a blank queryset
                queryset = queryset.none()
            else:
                queryset = queryset.filter(created__gte=date_from)

        # filter on date_to (yyyy-mm-dd)
        date_to = self.request.query_params.get("date_to", None)
        if date_to is not None:
            try:
                date_to = datetime.datetime.strptime(date_to, "%Y-%m-%d")
                date_to = timezone.make_aware(date_to, tz)
            except ValueError:
                # return a blank queryset
                queryset = queryset.none()
            else:
                queryset = queryset.filter(created__lte=date_to)

        return queryset


class FSMSnapshotViewMixin:
    def pre_transition(self, instance, action):
        super().pre_transition(instance, action)
        self._pre_save = create_dict_with_relations(instance)

    def post_transition(self, instance, action):
        super().post_transition(instance, action)
        create_snapshot(instance, self._pre_save, self.request.user)


src/unicef_snapshot/models.py
from django.conf import settings
from django.contrib.contenttypes.fields import GenericForeignKey
from django.contrib.contenttypes.models import ContentType
from django.core.exceptions import FieldDoesNotExist
from django.db import models
from django.utils.translation import gettext as _
from model_utils.models import TimeStampedModel


class Activity(TimeStampedModel):
    CREATE = "create"
    UPDATE = "update"
    ACTION_CHOICES = (
        (CREATE, _("Create")),
        (UPDATE, _("Update")),
    )

    target_content_type = models.ForeignKey(
        ContentType,
        related_name="activity",
        on_delete=models.CASCADE,
        db_index=True,
        verbose_name=_("Content Type"),
    )
    target_object_id = models.CharField(
        max_length=255,
        db_index=True,
        verbose_name=_("Target Object ID"),
    )
    target = GenericForeignKey("target_content_type", "target_object_id")
    action = models.CharField(
        verbose_name=_("Action"),
        max_length=50,
        choices=ACTION_CHOICES,
    )
    by_user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        verbose_name=_("By User"),
        on_delete=models.CASCADE,
    )
    data = models.JSONField(verbose_name=_("Data"))
    change = models.JSONField(verbose_name=_("Change"), blank=True)

    class Meta:
        ordering = ["-created"]
        verbose_name_plural = _("Activities")

    def __str__(self):
        return "{} {} {}".format(self.by_user, self.action, self.target)

    def by_user_display(self):
        by_user = str(self.by_user)
        if not by_user.strip():
            by_user = self.by_user.email
        return by_user

    def get_display(self):
        if self.action == Activity.CREATE:
            return _("Created")
        elif self.action == Activity.UPDATE:
            titles = []
            try:
                for field in self.change.keys():
                    try:
                        target_field = self.target._meta.get_field(field)
                        if hasattr(target_field, "verbose_name"):
                            field_name = target_field.verbose_name
                        else:
                            field_name = field
                        titles.append(str(field_name))
                    except FieldDoesNotExist:
                        continue
            except AttributeError:
                titles.append("unknown")

            return _("Changed {}").format(", ".join(titles))
        else:
            raise NotImplementedError


src/unicef_snapshot/migrations/0001_initial.py
# Generated by Django 1.10.8 on 2018-03-26 16:05

import django.db.models.deletion
import django.utils.timezone
import model_utils.fields
from django.conf import settings
from django.db import migrations, models


class Migration(migrations.Migration):
    initial = True

    dependencies = [
        ("contenttypes", "0002_remove_content_type_name"),
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name="Activity",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                (
                    "created",
                    model_utils.fields.AutoCreatedField(
                        default=django.utils.timezone.now,
                        editable=False,
                        verbose_name="created",
                    ),
                ),
                (
                    "modified",
                    model_utils.fields.AutoLastModifiedField(
                        default=django.utils.timezone.now,
                        editable=False,
                        verbose_name="modified",
                    ),
                ),
                (
                    "target_object_id",
                    models.CharField(
                        db_index=True, max_length=255, verbose_name="Target Object ID"
                    ),
                ),
                (
                    "action",
                    models.CharField(
                        choices=[("create", "Create"), ("update", "Update")],
                        max_length=50,
                        verbose_name="Action",
                    ),
                ),
                ("data", models.JSONField(verbose_name="Data")),
                ("change", models.JSONField(blank=True, verbose_name="Change")),
                (
                    "by_user",
                    models.ForeignKey(
                        on_delete=django.db.models.deletion.CASCADE,
                        to=settings.AUTH_USER_MODEL,
                        verbose_name="By User",
                    ),
                ),
                (
                    "target_content_type",
                    models.ForeignKey(
                        on_delete=django.db.models.deletion.CASCADE,
                        related_name="activity",
                        to="contenttypes.ContentType",
                        verbose_name="Content Type",
                    ),
                ),
            ],
            options={
                "ordering": ["-created"],
                "verbose_name_plural": "Activities",
            },
        ),
    ]
